File: 백준/Silver/2512. 예산/예산.py
import sys

class Solution:
    def __init__(self):
        self.input = sys.stdin.readline
        self.output = sys.stdout.write
    
    def determination(self, nums, target, mid):
        res = 0
        for n in nums:
                res += min(n, mid)
        return res <= target

    def answer(self):
        n = int(self.input())
        nums = [int(i) for i in self.input().split()]
        target = int(self.input())
        left, right, ans = 0, max(nums), 0
        
        while left <= right:
            mid = (left + right) // 2
            if self.determination(nums, target, mid):
                ans = mid
                left = mid + 1
                continue

            right = mid - 1

        self.output(f'{ans}\n')

s = Solution()
s.answer()

File: 백준/Silver/2512. 예산/예산.java
import java.io.*;
import java.util.*;

public class Main {

    private static boolean calFunc(int[] nums, int mid, int target ){
        int temp_res = 0;
        for(int num : nums) temp_res += (num < mid) ? num : mid;

        return temp_res <= target;
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        String[] temp = br.readLine().split(" ");
        int[] nums = new int[n];

        for(int i = 0; i < n ; i++) nums[i] = Integer.parseInt(temp[i]);      

        int m = Integer.parseInt(br.readLine());

        int left = 0;
        int right = Arrays.stream(nums).max().getAsInt();
        int mid = 0;
        int answer = 0;

        while (left <= right) {
            mid = (left + right) / 2;

            if(calFunc(nums, mid, m)){
                answer = mid;
                left = mid + 1;
                continue;
            }
            right = mid - 1;
        }

        System.out.println(answer);
        br.close();
    }
}